File: src/stores/groups.ts
import { defineStore } from 'pinia';
import { ref } from 'vue';
import { api } from 'boot/axios';

export interface Group {
  id: string;
  name: string;
  api_key: string;
  failover_status_codes: number[];
  is_active: boolean;
  ttft_timeout_ms: number | null;
  count_tokens_server_id: string | null;
  count_tokens_model_mappings: Record<string, string>;
  created_at: string;
  updated_at: string;
}

export interface Model {
  id: string;
  name: string;
  input_1m_usd: number | null;
  output_1m_usd: number | null;
  cache_write_1m_usd: number | null;
  cache_read_1m_usd: number | null;
  created_at: string;
}

export interface GroupServerDetail {
  server_id: string;
  short_id: number;
  server_name: string;
  base_url: string | null;
  api_key: string | null;
  priority: number;
  model_mappings: Record<string, string>;
  is_enabled: boolean;
  cb_max_failures: number | null;
  cb_window_seconds: number | null;
  cb_cooldown_seconds: number | null;
  rate_input: number | null;
  rate_output: number | null;
  rate_cache_write: number | null;
  rate_cache_read: number | null;
  max_requests: number | null;
  rate_window_seconds: number | null;
  normalize_cache_read: boolean;
  max_input_tokens: number | null;
  min_input_tokens: number | null;
  supported_models: string[];
  password_hash: string | null;
  active_hours_start: string | null;
  active_hours_end: string | null;
  active_hours_timezone: string | null;
  retry_status_codes: number[] | null;
  retry_count: number | null;
  retry_delay_seconds: number | null;
}

export interface GroupWithServers extends Group {
  servers: GroupServerDetail[];
  allowed_models: Model[];
}

interface PaginatedResponse {
  data: Group[];
  total: number;
  page: number;
  total_pages: number;
}

export interface CircuitStatus {
  server_id: string;
  is_open: boolean;
  remaining_seconds: number;
}

export interface TtftDataPoint {
  created_at: string;
  ttft_ms: number | null;
  timed_out: boolean;
}

export interface ServerTtftStats {
  server_id: string;
  server_name: string;
  avg_ttft_ms: number | null;
  p50_ttft_ms: number | null;
  p95_ttft_ms: number | null;
  timeout_count: number;
  total_count: number;
  data_points: TtftDataPoint[];
}

export interface TtftStatsResponse {
  servers: ServerTtftStats[];
}

export interface ServerTokenUsage {
  server_id: string;
  server_name: string;
  model: string | null;
  total_input_tokens: number;
  total_output_tokens: number;
  total_cache_creation_tokens: number;
  total_cache_read_tokens: number;
  request_count: number;
  cost_usd: number | null;
}

export interface TokenUsageStats {
  servers: ServerTokenUsage[];
}

export interface KeyTokenUsage {
  group_key_id: string | null;
  key_name: string | null;
  api_key: string | null;
  created_at: string | null;
  total_input_tokens: number;
  total_output_tokens: number;
  total_cache_creation_tokens: number;
  total_cache_read_tokens: number;
  request_count: number;
  cost_usd: number | null;
}

export interface KeyUsageResponse {
  keys: KeyTokenUsage[];
}

export interface SpamResult {
  group_key_id: string;
  api_key: string;
  key_name: string;
  spam_type: 'low_token' | 'duplicate_request';
  request_count: number;
  peak_rpm: number;
  detected_at: string;
}

export interface GroupKey {
  id: string;
  group_id: string;
  api_key: string;
  name: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

interface PaginatedGroupKeys {
  data: GroupKey[];
  total: number;
  page: number;
  total_pages: number;
}

export const useGroupsStore = defineStore('groups', () => {
  const groups = ref<Group[]>([]);
  const total = ref(0);
  const totalPages = ref(0);
  const loading = ref(false);

  async function fetchGroups(params?: {
    page?: number; limit?: number; search?: string;
    is_active?: boolean; server_id?: string;
  }) {
    loading.value = true;
    try {
      const { data } = await api.get<PaginatedResponse>('/api/admin/groups', { params });
      groups.value = data.data;
      total.value = data.total;
      totalPages.value = data.total_pages;
      return data;
    } finally {
      loading.value = false;
    }
  }

  async function getGroup(id: string) {
    const { data } = await api.get<GroupWithServers>(`/api/admin/groups/${id}`);
    return data;
  }

  async function createGroup(input: { name: string; failover_status_codes?: number[] }) {
    const { data } = await api.post<Group>('/api/admin/groups', input);
    return data;
  }

  async function updateGroup(id: string, input: { name?: string; failover_status_codes?: number[]; is_active?: boolean; ttft_timeout_ms?: number | null; count_tokens_server_id?: string | null; count_tokens_model_mappings?: Record<string, string> }) {
    const { data } = await api.put<Group>(`/api/admin/groups/${id}`, input);
    return data;
  }

  async function deleteGroup(id: string) {
    await api.delete(`/api/admin/groups/${id}`);
  }

  async function regenerateKey(id: string) {
    const { data } = await api.post<Group>(`/api/admin/groups/${id}/regenerate-key`);
    return data;
  }

  async function bulkActivate(ids: string[]) {
    await api.post('/api/admin/groups/bulk/activate', { ids });
  }

  async function bulkDeactivate(ids: string[]) {
    await api.post('/api/admin/groups/bulk/deactivate', { ids });
  }

  async function bulkDelete(ids: string[]) {
    await api.post('/api/admin/groups/bulk/delete', { ids });
  }

  async function bulkAssignServer(group_ids: string[], server_id: string, priority: number, model_mappings?: Record<string, string>) {
    await api.post('/api/admin/groups/bulk/assign-server', { group_ids, server_id, priority, model_mappings });
  }

  // Group-server assignment actions
  async function assignServer(groupId: string, input: { server_id: string; priority: number; model_mappings?: Record<string, string> }) {
    await api.post(`/api/admin/groups/${groupId}/servers`, input);
  }

  async function updateAssignment(
    groupId: string,
    serverId: string,
    input: {
      priority?: number;
      model_mappings?: Record<string, string>;
      is_enabled?: boolean;
      cb_max_failures?: number | null;
      cb_window_seconds?: number | null;
      cb_cooldown_seconds?: number | null;
      rate_input?: number | null;
      rate_output?: number | null;
      rate_cache_write?: number | null;
      rate_cache_read?: number | null;
      max_requests?: number | null;
      rate_window_seconds?: number | null;
      normalize_cache_read?: boolean;
      max_input_tokens?: number | null;
      min_input_tokens?: number | null;
      supported_models?: string[];
      active_hours_start?: string | null;
      active_hours_end?: string | null;
      active_hours_timezone?: string | null;
      retry_status_codes?: number[] | null;
      retry_count?: number | null;
      retry_delay_seconds?: number | null;
    },
  ) {
    await api.put(`/api/admin/groups/${groupId}/servers/${serverId}`, input);
  }

  async function removeServer(groupId: string, serverId: string) {
    await api.delete(`/api/admin/groups/${groupId}/servers/${serverId}`);
  }

  async function reorderServers(groupId: string, serverIds: string[]) {
    await api.put(`/api/admin/groups/${groupId}/servers/reorder`, { server_ids: serverIds });
  }

  async function fetchTtftStats(groupId: string, range?: { start: string; end: string }, groupKeyId?: string) {
    const params: Record<string, string> = { group_id: groupId };
    if (range) {
      params.start = range.start;
      params.end = range.end;
    } else {
      params.period = '1h';
    }
    if (groupKeyId) {
      params.group_key_id = groupKeyId;
    }
    const { data } = await api.get<TtftStatsResponse>('/api/admin/ttft-stats', { params });
    return data;
  }

  async function fetchCircuitStatus(groupId: string) {
    const { data } = await api.get<CircuitStatus[]>(
      `/api/admin/groups/${groupId}/circuit-status`,
    );
    return data;
  }

  async function fetchTokenUsageStats(
    groupId: string,
    params?: { start?: string; end?: string; period?: string; is_dynamic_key?: boolean; key_hash?: string; group_key_id?: string },
  ) {
    const qp: Record<string, string> = { group_id: groupId };
    if (params?.start) qp.start = params.start;
    if (params?.end) qp.end = params.end;
    if (params?.period) qp.period = params.period;
    if (params?.is_dynamic_key !== undefined) qp.is_dynamic_key = String(params.is_dynamic_key);
    if (params?.key_hash) qp.key_hash = params.key_hash;
    if (params?.group_key_id) qp.group_key_id = params.group_key_id;
    const { data } = await api.get<TokenUsageStats>('/api/admin/token-usage', { params: qp });
    return data;
  }

  async function fetchTokenUsageByKey(
    groupId: string,
    params?: { start?: string; end?: string; period?: string },
  ) {
    const qp: Record<string, string> = { group_id: groupId };
    if (params?.start) qp.start = params.start;
    if (params?.end) qp.end = params.end;
    if (params?.period) qp.period = params.period;
    const { data } = await api.get<KeyUsageResponse>('/api/admin/token-usage/by-key', { params: qp });
    return data;
  }

  async function fetchGroupKeys(
    groupId: string,
    params?: { page?: number; limit?: number; search?: string },
  ) {
    const { data } = await api.get<PaginatedGroupKeys>(`/api/admin/groups/${groupId}/keys`, { params });
    return data;
  }

  async function createGroupKey(groupId: string, name: string) {
    const { data } = await api.post<GroupKey>(`/api/admin/groups/${groupId}/keys`, { name });
    return data;
  }

  async function updateGroupKey(groupId: string, keyId: string, input: { name?: string; is_active?: boolean }) {
    const { data } = await api.patch<GroupKey>(`/api/admin/groups/${groupId}/keys/${keyId}`, input);
    return data;
  }

  async function regenerateGroupKey(groupId: string, keyId: string) {
    const { data } = await api.post<GroupKey>(`/api/admin/groups/${groupId}/keys/${keyId}/regenerate`);
    return data;
  }

  async function bulkCreateGroupKeys(
    groupId: string,
    input: { count: number; plan_id: string; name_prefix?: string },
  ) {
    const { data } = await api.post<GroupKey[]>(`/api/admin/groups/${groupId}/keys/bulk`, input);
    return data;
  }

  async function fetchKeyUsage(
    groupId: string,
    groupKeyId: string,
    params?: { period?: string; start?: string; end?: string },
  ) {
    return fetchTokenUsageStats(groupId, { ...params, group_key_id: groupKeyId });
  }

  // Group allowed models
  async function fetchGroupAllowedModels(groupId: string) {
    const { data } = await api.get<Model[]>(`/api/admin/groups/${groupId}/allowed-models`);
    return data;
  }

  async function addGroupAllowedModel(groupId: string, input: { model_id?: string; name?: string }) {
    const { data } = await api.post<Model>(`/api/admin/groups/${groupId}/allowed-models`, input);
    return data;
  }

  async function removeGroupAllowedModel(groupId: string, modelId: string) {
    await api.delete(`/api/admin/groups/${groupId}/allowed-models/${modelId}`);
  }

  // Key allowed models
  async function fetchKeyAllowedModels(groupId: string, keyId: string) {
    const { data } = await api.get<Model[]>(`/api/admin/groups/${groupId}/keys/${keyId}/allowed-models`);
    return data;
  }

  async function addKeyAllowedModel(groupId: string, keyId: string, modelId: string) {
    const { data } = await api.post<Model>(`/api/admin/groups/${groupId}/keys/${keyId}/allowed-models`, { model_id: modelId });
    return data;
  }

  async function removeKeyAllowedModel(groupId: string, keyId: string, modelId: string) {
    await api.delete(`/api/admin/groups/${groupId}/keys/${keyId}/allowed-models/${modelId}`);
  }

  // Key server assignments
  async function fetchKeyServers(groupId: string, keyId: string) {
    const { data } = await api.get<GroupServerDetail[]>(`/api/admin/groups/${groupId}/keys/${keyId}/servers`);
    return data;
  }

  async function assignKeyServer(groupId: string, keyId: string, serverId: string) {
    const { data } = await api.post<GroupServerDetail[]>(`/api/admin/groups/${groupId}/keys/${keyId}/servers`, { server_ids: [serverId] });
    return data;
  }

  async function removeKeyServer(groupId: string, keyId: string, serverId: string) {
    await api.delete(`/api/admin/groups/${groupId}/keys/${keyId}/servers/${serverId}`);
  }

  async function fetchSpamDetection(groupId: string, params?: { page?: number; limit?: number }) {
    const { data } = await api.get<{
      data: SpamResult[];
      total: number;
      page: number;
      total_pages: number;
    }>('/api/admin/spam-detection', { params: { group_id: groupId, ...params } });
    return data;
  }

  // User agent management
  async function fetchGroupUserAgents(groupId: string) {
    const { data } = await api.get<{ user_agent: string; first_seen_at: string }[]>(
      `/api/admin/groups/${groupId}/user-agents`,
    );
    return data;
  }

  async function fetchGroupBlockedUserAgents(groupId: string) {
    const { data } = await api.get<{ user_agent: string; created_at: string }[]>(
      `/api/admin/groups/${groupId}/user-agents/blocked`,
    );
    return data;
  }

  async function addGroupBlockedUserAgent(groupId: string, userAgent: string) {
    await api.post(`/api/admin/groups/${groupId}/user-agents/blocked`, { user_agent: userAgent });
  }

  async function removeGroupBlockedUserAgent(groupId: string, userAgent: string) {
    await api.delete(`/api/admin/groups/${groupId}/user-agents/blocked`, {
      data: { user_agent: userAgent },
    });
  }

  return {
    groups, total, totalPages, loading,
    fetchGroups, getGroup, createGroup, updateGroup, deleteGroup, regenerateKey,
    bulkActivate, bulkDeactivate, bulkDelete, bulkAssignServer,
    assignServer, updateAssignment, removeServer, reorderServers,
    fetchTtftStats, fetchCircuitStatus, fetchTokenUsageStats, fetchTokenUsageByKey,
    fetchGroupKeys, createGroupKey, updateGroupKey, regenerateGroupKey, bulkCreateGroupKeys, fetchKeyUsage,
    fetchGroupAllowedModels, addGroupAllowedModel, removeGroupAllowedModel,
    fetchKeyAllowedModels, addKeyAllowedModel, removeKeyAllowedModel,
    fetchKeyServers, assignKeyServer, removeKeyServer,
    fetchSpamDetection,
    fetchGroupUserAgents, fetchGroupBlockedUserAgents, addGroupBlockedUserAgent, removeGroupBlockedUserAgent,
  };
});

File: viber-router-api/src/routes/admin/groups.rs
use axum::{
    Json, Router,
    extract::{Path, Query, State},
    http::StatusCode,
    routing::{get, post},
};
use serde::Deserialize;
use uuid::Uuid;

use crate::models::{
    AdminGroupServerDetail, CreateGroup, Group, GroupListItem, GroupWithServers, Model, PaginatedResponse, UpdateGroup,
    generate_api_key,
};
use crate::routes::AppState;

#[derive(Debug, Deserialize)]
pub struct ListParams {
    pub page: Option<i64>,
    pub limit: Option<i64>,
    pub search: Option<String>,
    pub is_active: Option<bool>,
    pub server_id: Option<Uuid>,
    pub order: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct BulkIds {
    pub ids: Vec<Uuid>,
}

#[derive(Debug, Deserialize)]
pub struct BulkAssignServer {
    pub group_ids: Vec<Uuid>,
    pub server_id: Uuid,
    pub priority: i32,
    pub model_mappings: Option<serde_json::Value>,
}

type ApiError = (StatusCode, Json<serde_json::Value>);

fn err(status: StatusCode, msg: &str) -> ApiError {
    (status, Json(serde_json::json!({"error": msg})))
}

fn internal(e: impl std::fmt::Display) -> ApiError {
    err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string())
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", get(list_groups).post(create_group))
        .route("/{id}", get(get_group).put(update_group).delete(delete_group))
        .route("/{id}/regenerate-key", post(regenerate_key))
        .route("/{id}/circuit-status", get(circuit_status))
        .route("/bulk/activate", post(bulk_activate))
        .route("/bulk/deactivate", post(bulk_deactivate))
        .route("/bulk/delete", post(bulk_delete))
        .route("/bulk/assign-server", post(bulk_assign_server))
        .nest("/{group_id}/servers", super::group_servers::router())
        .nest("/{group_id}/keys", super::group_keys::router())
        .nest("/{group_id}/allowed-models", super::group_allowed_models::router())
        .nest("/{group_id}/uptime", super::uptime::router())
        .nest("/{group_id}/user-agents", super::group_user_agents::router())
}

async fn create_group(
    State(state): State<AppState>,
    Json(input): Json<CreateGroup>,
) -> Result<(StatusCode, Json<Group>), ApiError> {
    let api_key = generate_api_key();
    let codes = input
        .failover_status_codes
        .unwrap_or_else(|| vec![429, 500, 502, 503]);
    let codes_json = serde_json::to_value(&codes).map_err(internal)?;

    let group = sqlx::query_as::<_, Group>(
        "INSERT INTO groups (name, api_key, failover_status_codes) VALUES ($1, $2, $3) RETURNING *",
    )
    .bind(&input.name)
    .bind(&api_key)
    .bind(&codes_json)
    .fetch_one(&state.db)
    .await
    .map_err(internal)?;

    Ok((StatusCode::CREATED, Json(group)))
}

async fn list_groups(
    State(state): State<AppState>,
    Query(params): Query<ListParams>,
) -> Result<Json<PaginatedResponse<GroupListItem>>, ApiError> {
    let page = params.page.unwrap_or(1).max(1);
    let limit = params.limit.unwrap_or(20).clamp(1, 100);
    let offset = (page - 1) * limit;
    let order = if params.order.as_deref() == Some("asc") { "ASC" } else { "DESC" };

    let mut conditions = vec!["1=1".to_string()];
    let mut bind_idx = 0u32;

    // Build dynamic query
    let search_pattern = params.search.as_ref().map(|s| format!("%{s}%"));
    if search_pattern.is_some() {
        bind_idx += 1;
        conditions.push(format!("g.name ILIKE ${bind_idx}"));
    }
    if params.is_active.is_some() {
        bind_idx += 1;
        conditions.push(format!("g.is_active = ${bind_idx}"));
    }
    if params.server_id.is_some() {
        bind_idx += 1;
        conditions.push(format!(
            "EXISTS (SELECT 1 FROM group_servers gs WHERE gs.group_id = g.id AND gs.server_id = ${bind_idx})"
        ));
    }

    let where_clause = conditions.join(" AND ");
    let count_sql = format!("SELECT COUNT(*) FROM groups g WHERE {where_clause}");
    let data_sql = format!(
        "SELECT g.*, COALESCE((SELECT COUNT(*) FROM group_servers gs2 WHERE gs2.group_id = g.id), 0) as servers_count \
         FROM groups g WHERE {where_clause} ORDER BY g.created_at {order} LIMIT ${} OFFSET ${}",
        bind_idx + 1,
        bind_idx + 2
    );

    // Use raw query builder to bind dynamically
    let mut count_query = sqlx::query_as::<_, (i64,)>(&count_sql);
    let mut data_query = sqlx::query_as::<_, GroupListItem>(&data_sql);

    if let Some(ref pattern) = search_pattern {
        count_query = count_query.bind(pattern);
        data_query = data_query.bind(pattern);
    }
    if let Some(is_active) = params.is_active {
        count_query = count_query.bind(is_active);
        data_query = data_query.bind(is_active);
    }
    if let Some(server_id) = params.server_id {
        count_query = count_query.bind(server_id);
        data_query = data_query.bind(server_id);
    }

    data_query = data_query.bind(limit).bind(offset);

    let (total,) = count_query.fetch_one(&state.db).await.map_err(internal)?;
    let groups = data_query.fetch_all(&state.db).await.map_err(internal)?;
    let total_pages = (total as f64 / limit as f64).ceil() as i64;

    Ok(Json(PaginatedResponse { data: groups, total, page, total_pages }))
}

async fn get_group(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<GroupWithServers>, ApiError> {
    let group = sqlx::query_as::<_, Group>("SELECT * FROM groups WHERE id = $1")
        .bind(id)
        .fetch_optional(&state.db)
        .await
        .map_err(internal)?
        .ok_or_else(|| err(StatusCode::NOT_FOUND, "Group not found"))?;

    let unlocked = state.unlocked_servers.read().await;

    let raw_servers: Vec<AdminGroupServerDetail> = sqlx::query_as::<_, AdminGroupServerDetail>(
        "SELECT gs.server_id, s.short_id, s.name as server_name, s.base_url, s.api_key, gs.priority, gs.model_mappings, gs.is_enabled, \
         gs.cb_max_failures, gs.cb_window_seconds, gs.cb_cooldown_seconds, \
         gs.rate_input, gs.rate_output, gs.rate_cache_write, gs.rate_cache_read, \
         gs.max_requests, gs.rate_window_seconds, gs.normalize_cache_read, gs.max_input_tokens, gs.min_input_tokens, gs.supported_models, s.password_hash \
         FROM group_servers gs JOIN servers s ON s.id = gs.server_id \
         WHERE gs.group_id = $1 ORDER BY gs.priority",
    )
    .bind(id)
    .fetch_all(&state.db)
    .await
    .map_err(internal)?;

    // Mask credentials for locked servers
    let servers: Vec<AdminGroupServerDetail> = raw_servers
        .into_iter()
        .map(|mut s| {
            let is_locked = s.password_hash.is_some() && !unlocked.contains(&s.server_id);
            if is_locked {
                s.base_url = None;
                s.api_key = None;
            }
            s
        })
        .collect();

    let allowed_models = sqlx::query_as::<_, Model>(
        "SELECT m.* FROM models m \
         JOIN group_allowed_models gam ON m.id = gam.model_id \
         WHERE gam.group_id = $1 ORDER BY m.name",
    )
    .bind(id)
    .fetch_all(&state.db)
    .await
    .map_err(internal)?;

    Ok(Json(GroupWithServers { group, servers, allowed_models }))
}

#[derive(Debug, serde::Serialize)]
struct CircuitStatusEntry {
    server_id: Uuid,
    is_open: bool,
    remaining_seconds: i64,
}

async fn circuit_status(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<Vec<CircuitStatusEntry>>, ApiError> {
    // Get all servers with CB configured for this group
    let rows = sqlx::query_as::<_, (Uuid,)>(
        "SELECT server_id FROM group_servers WHERE group_id = $1 AND cb_max_failures IS NOT NULL",
    )
    .bind(id)
    .fetch_all(&state.db)
    .await
    .map_err(internal)?;

    let mut entries = Vec::new();
    if let Ok(mut conn) = state.redis.get().await {
        for (server_id,) in rows {
            let key = format!("cb:open:{id}:{server_id}");
            let ttl: i64 = deadpool_redis::redis::cmd("TTL")
                .arg(&key)
                .query_async(&mut conn)
                .await
                .unwrap_or(-2);
            // TTL: -2 = key doesn't exist, -1 = no expiry, >0 = remaining seconds
            let is_open = ttl > 0;
            entries.push(CircuitStatusEntry {
                server_id,
                is_open,
                remaining_seconds: if is_open { ttl } else { 0 },
            });
        }
    }

    Ok(Json(entries))
}

async fn update_group(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(input): Json<UpdateGroup>,
) -> Result<Json<Group>, ApiError> {
    let codes_json = input
        .failover_status_codes
        .as_ref()
        .map(serde_json::to_value)
        .transpose()
        .map_err(internal)?;

    // For ttft_timeout_ms: None means "don't change", Some(None) means "set to NULL", Some(Some(v)) means "set to v"
    let (update_ttft, ttft_val) = match input.ttft_timeout_ms {
        Some(v) => (true, v),
        None => (false, None),
    };

    // Same pattern for count_tokens_server_id
    let (update_ct_server, ct_server_val) = match input.count_tokens_server_id {
        Some(v) => (true, v),
        None => (false, None),
    };

    // For count_tokens_model_mappings: None means "don't change", Some(v) means "set to v"
    let update_ct_mappings = input.count_tokens_model_mappings.is_some();
    let ct_mappings_val = input.count_tokens_model_mappings.unwrap_or_default();

    let group = sqlx::query_as::<_, Group>(
        "UPDATE groups SET \
         name = COALESCE($1, name), \
         failover_status_codes = COALESCE($2, failover_status_codes), \
         is_active = COALESCE($3, is_active), \
         ttft_timeout_ms = CASE WHEN $5 THEN $6 ELSE ttft_timeout_ms END, \
         count_tokens_server_id = CASE WHEN $7 THEN $8 ELSE count_tokens_server_id END, \
         count_tokens_model_mappings = CASE WHEN $9 THEN $10 ELSE count_tokens_model_mappings END, \
         updated_at = now() \
         WHERE id = $4 RETURNING *",
    )
    .bind(&input.name)
    .bind(&codes_json)
    .bind(input.is_active)
    .bind(id)
    .bind(update_ttft)
    .bind(ttft_val)
    .bind(update_ct_server)
    .bind(ct_server_val)
    .bind(update_ct_mappings)
    .bind(&ct_mappings_val)
    .fetch_optional(&state.db)
    .await
    .map_err(internal)?
    .ok_or_else(|| err(StatusCode::NOT_FOUND, "Group not found"))?;

    crate::cache::invalidate_group_all_keys(&state.redis, &state.db, id).await;
    Ok(Json(group))
}

async fn delete_group(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<StatusCode, ApiError> {
    // Verify group exists
    let exists = sqlx::query_scalar::<_, bool>("SELECT EXISTS(SELECT 1 FROM groups WHERE id = $1)")
        .bind(id)
        .fetch_one(&state.db)
        .await
        .map_err(internal)?;
    if !exists {
        return Err(err(StatusCode::NOT_FOUND, "Group not found"));
    }

    // Invalidate before delete (sub-keys will be cascade-deleted)
    crate::cache::invalidate_group_all_keys(&state.redis, &state.db, id).await;

    sqlx::query("DELETE FROM groups WHERE id = $1")
        .bind(id)
        .execute(&state.db)
        .await
        .map_err(internal)?;

    Ok(StatusCode::NO_CONTENT)
}

async fn regenerate_key(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<Group>, ApiError> {
    let old_group = sqlx::query_as::<_, Group>("SELECT * FROM groups WHERE id = $1")
        .bind(id)
        .fetch_optional(&state.db)
        .await
        .map_err(internal)?
        .ok_or_else(|| err(StatusCode::NOT_FOUND, "Group not found"))?;

    let new_key = generate_api_key();
    let group = sqlx::query_as::<_, Group>(
        "UPDATE groups SET api_key = $1, updated_at = now() WHERE id = $2 RETURNING *",
    )
    .bind(&new_key)
    .bind(id)
    .fetch_one(&state.db)
    .await
    .map_err(internal)?;

    // Invalidate old master key + all sub-keys (sub-keys cache the same GroupConfig)
    crate::cache::invalidate_group_config(&state.redis, &old_group.api_key).await;
    crate::cache::invalidate_group_all_keys(&state.redis, &state.db, id).await;
    Ok(Json(group))
}

async fn bulk_activate(
    State(state): State<AppState>,
    Json(input): Json<BulkIds>,
) -> Result<StatusCode, ApiError> {
    let groups = sqlx::query_as::<_, Group>(
        "UPDATE groups SET is_active = true, updated_at = now() WHERE id = ANY($1) RETURNING *",
    )
    .bind(&input.ids)
    .fetch_all(&state.db)
    .await
    .map_err(internal)?;

    for g in &groups {
        crate::cache::invalidate_group_all_keys(&state.redis, &state.db, g.id).await;
    }
    Ok(StatusCode::OK)
}

async fn bulk_deactivate(
    State(state): State<AppState>,
    Json(input): Json<BulkIds>,
) -> Result<StatusCode, ApiError> {
    let groups = sqlx::query_as::<_, Group>(
        "UPDATE groups SET is_active = false, updated_at = now() WHERE id = ANY($1) RETURNING *",
    )
    .bind(&input.ids)
    .fetch_all(&state.db)
    .await
    .map_err(internal)?;

    for g in &groups {
        crate::cache::invalidate_group_all_keys(&state.redis, &state.db, g.id).await;
    }
    Ok(StatusCode::OK)
}

async fn bulk_delete(
    State(state): State<AppState>,
    Json(input): Json<BulkIds>,
) -> Result<StatusCode, ApiError> {
    let groups = sqlx::query_as::<_, Group>(
        "SELECT * FROM groups WHERE id = ANY($1)",
    )
    .bind(&input.ids)
    .fetch_all(&state.db)
    .await
    .map_err(internal)?;

    // Invalidate before delete
    for g in &groups {
        crate::cache::invalidate_group_all_keys(&state.redis, &state.db, g.id).await;
    }

    sqlx::query("DELETE FROM groups WHERE id = ANY($1)")
        .bind(&input.ids)
        .execute(&state.db)
        .await
        .map_err(internal)?;

    Ok(StatusCode::OK)
}

async fn bulk_assign_server(
    State(state): State<AppState>,
    Json(input): Json<BulkAssignServer>,
) -> Result<StatusCode, ApiError> {
    let mappings = input.model_mappings.unwrap_or(serde_json::json!({}));

    for group_id in &input.group_ids {
        sqlx::query(
            "INSERT INTO group_servers (group_id, server_id, priority, model_mappings, is_enabled) \
             VALUES ($1, $2, $3, $4, true) \
             ON CONFLICT (group_id, server_id) DO UPDATE SET priority = $3, model_mappings = $4, is_enabled = true",
        )
        .bind(group_id)
        .bind(input.server_id)
        .bind(input.priority)
        .bind(&mappings)
        .execute(&state.db)
        .await
        .map_err(internal)?;
    }

    // Invalidate cache for all affected groups
    for group_id in &input.group_ids {
        crate::cache::invalidate_group_all_keys(&state.redis, &state.db, *group_id).await;
    }
    Ok(StatusCode::OK)
}
